File: backend/src/main/java/com/inventory/service/ProductService.java
package com.inventory.service;

import com.inventory.dto.ProductDTO;
import com.inventory.dto.ProductMateriaPrimaDTO;
import com.inventory.entity.Product;
import com.inventory.entity.ProductMateriaPrima;
import com.inventory.entity.MateriaPrima;
import com.inventory.exception.ResourceNotFoundException;
import com.inventory.repository.ProductMateriaPrimaRepository;
import com.inventory.repository.ProductRepository;
import com.inventory.repository.MateriaPrimaRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ProductService {
    
    private final ProductRepository productRepository;
    private final MateriaPrimaRepository materiaPrimaRepository;
    private final ProductMateriaPrimaRepository productMateriaPrimaRepository;
    
    @Transactional(readOnly = true)
    public List<ProductDTO> findAll() {
        return productRepository.findAllWithMateriaPrimas().stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }
    
    @Transactional(readOnly = true)
    public ProductDTO findById(Long id) {
        Long safeId = Objects.requireNonNull(id, "id is required");
        Product product = productRepository.findByIdWithMateriaPrimas(safeId);
        if (product == null) {
            throw new ResourceNotFoundException("Product not found with id: " + safeId);
        }
        return convertToDTO(product);
    }
    
    @Transactional
    public ProductDTO create(ProductDTO productDTO) {
        Product product = new Product();
        product.setName(productDTO.getName());
        product.setValue(productDTO.getValue());
        product.setCode(generateTemporaryCode());

        Product savedProduct = productRepository.save(product);
        savedProduct.setCode(generateProductCode(savedProduct.getId()));
        savedProduct = productRepository.save(savedProduct);
        
        if (productDTO.getMateriasPrimas() != null && !productDTO.getMateriasPrimas().isEmpty()) {
            updateProductMateriasPrimas(savedProduct, productDTO.getMateriasPrimas());
        }
        
        return convertToDTO(productRepository.findByIdWithMateriaPrimas(savedProduct.getId()));
    }
    
    @Transactional
    public ProductDTO update(Long id, ProductDTO productDTO) {
        Long safeId = Objects.requireNonNull(id, "id is required");
        Product product = productRepository.findById(safeId)
            .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + safeId));
        
        product.setName(productDTO.getName());
        product.setValue(productDTO.getValue());
        
        if (productDTO.getMateriasPrimas() != null) {
            // Remove existing associations
            productMateriaPrimaRepository.deleteByProductId(safeId);
            product.getProductMateriaPrimas().clear();
            
            // Add new associations
            updateProductMateriasPrimas(product, productDTO.getMateriasPrimas());
        }
        
        productRepository.save(product);
        return convertToDTO(productRepository.findByIdWithMateriaPrimas(safeId));
    }
    
    @Transactional
    public void delete(Long id) {
        Long safeId = Objects.requireNonNull(id, "id is required");
        if (!productRepository.existsById(safeId)) {
            throw new ResourceNotFoundException("Product not found with id: " + safeId);
        }
        productRepository.deleteById(safeId);
    }
    
    private void updateProductMateriasPrimas(Product product, List<ProductMateriaPrimaDTO> materiasPrimasDTO) {
        for (ProductMateriaPrimaDTO dto : materiasPrimasDTO) {
            Long materiaPrimaId = Objects.requireNonNull(dto.getMateriaPrimaId(), "materiaPrimaId is required");
            MateriaPrima materiaPrima = materiaPrimaRepository.findById(materiaPrimaId)
                    .orElseThrow(() -> new ResourceNotFoundException("Materia prima not found with id: " + materiaPrimaId));
            
            ProductMateriaPrima productMateriaPrima = new ProductMateriaPrima();
            productMateriaPrima.setProduct(product);
            productMateriaPrima.setMateriaPrima(materiaPrima);
            productMateriaPrima.setQuantityRequired(dto.getQuantityRequired());
            
            product.getProductMateriaPrimas().add(productMateriaPrima);
        }
    }
    
    private ProductDTO convertToDTO(Product product) {
        ProductDTO dto = new ProductDTO();
        dto.setId(product.getId());
        dto.setName(product.getName());
        dto.setCode(product.getCode());
        dto.setValue(product.getValue());
        
        if (product.getProductMateriaPrimas() != null) {
            List<ProductMateriaPrimaDTO> materiasPrimasDTO = product.getProductMateriaPrimas().stream()
                    .map(prm -> {
                        ProductMateriaPrimaDTO rmDTO = new ProductMateriaPrimaDTO();
                        rmDTO.setId(prm.getId());
                        rmDTO.setMateriaPrimaId(prm.getMateriaPrima().getId());
                        rmDTO.setMateriaPrimaName(prm.getMateriaPrima().getName());
                        rmDTO.setQuantityRequired(prm.getQuantityRequired());
                        return rmDTO;
                    })
                    .collect(Collectors.toList());
            dto.setMateriasPrimas(materiasPrimasDTO);
        }
        
        return dto;
    }

    private String generateProductCode(Long id) {
        return String.format("PRD-%06d", id);
    }

    private String generateTemporaryCode() {
        return "TMP-" + java.util.UUID.randomUUID();
    }
}

File: backend/src/main/kotlin/com/inventory/service/ProductService.kt
package com.inventory.service

import com.inventory.dto.ProductDTO
import com.inventory.dto.ProductRawMaterialDTO
import com.inventory.entity.Product
import com.inventory.entity.ProductRawMaterial
import com.inventory.exception.ResourceNotFoundException
import com.inventory.repository.ProductRawMaterialRepository
import com.inventory.repository.ProductRepository
import com.inventory.repository.RawMaterialRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ProductService(
    private val productRepository: ProductRepository,
    private val rawMaterialRepository: RawMaterialRepository,
    private val productRawMaterialRepository: ProductRawMaterialRepository
) {
    
    @Transactional(readOnly = true)
    fun findAll(): List<ProductDTO> {
        return productRepository.findAllWithRawMaterials().map { convertToDTO(it) }
    }
    
    @Transactional(readOnly = true)
    fun findById(id: Long): ProductDTO {
        val product = productRepository.findByIdWithRawMaterials(id)
            ?: throw ResourceNotFoundException("Product not found with id: $id")
        return convertToDTO(product)
    }
    
    @Transactional
    fun create(productDTO: ProductDTO): ProductDTO {
        val product = Product(
            name = productDTO.name,
            value = productDTO.value
        )
        
        val savedProduct = productRepository.save(product)
        
        productDTO.rawMaterials?.takeIf { it.isNotEmpty() }?.let {
            updateProductRawMaterials(savedProduct, it)
        }
        
        return convertToDTO(productRepository.findByIdWithRawMaterials(savedProduct.id!!)!!)
    }
    
    @Transactional
    fun update(id: Long, productDTO: ProductDTO): ProductDTO {
        val product = productRepository.findById(id)
            .orElseThrow { ResourceNotFoundException("Product not found with id: $id") }
        
        product.name = productDTO.name
        product.value = productDTO.value
        
        productDTO.rawMaterials?.let {
            productRawMaterialRepository.deleteByProductId(id)
            product.productRawMaterials.clear()
            updateProductRawMaterials(product, it)
        }
        
        productRepository.save(product)
        return convertToDTO(productRepository.findByIdWithRawMaterials(id)!!)
    }
    
    @Transactional
    fun delete(id: Long) {
        if (!productRepository.existsById(id)) {
            throw ResourceNotFoundException("Product not found with id: $id")
        }
        productRepository.deleteById(id)
    }
    
    private fun updateProductRawMaterials(product: Product, rawMaterialsDTO: List<ProductRawMaterialDTO>) {
        rawMaterialsDTO.forEach { dto ->
            val rawMaterial = rawMaterialRepository.findById(dto.rawMaterialId!!)
                .orElseThrow { ResourceNotFoundException("Raw material not found with id: ${dto.rawMaterialId}") }
            
            val productRawMaterial = ProductRawMaterial(
                product = product,
                rawMaterial = rawMaterial,
                quantityRequired = dto.quantityRequired
            )
            
            product.productRawMaterials.add(productRawMaterial)
        }
    }
    
    private fun convertToDTO(product: Product): ProductDTO {
        return ProductDTO(
            id = product.id,
            name = product.name,
            value = product.value,
            rawMaterials = product.productRawMaterials.map { prm ->
                ProductRawMaterialDTO(
                    id = prm.id,
                    rawMaterialId = prm.rawMaterial?.id,
                    rawMaterialName = prm.rawMaterial?.name,
                    quantityRequired = prm.quantityRequired
                )
            }
        )
    }
}
